### ISSUE TITLE
Preserving semantic alignment and class discriminability after fine-tuning

### ISSUE THREAD

**Issue by buttisoft** [Sompote/sam3_lora#22]
Thank you for open-sourcing this project, great work!

I have been exploring your solution to improve SAM3 on specific objects while preserving its general recognition capabilities.

However, I noticed that after fine-tuning on my dataset, the model correctly identifies the target objects compared to the original SAM, but it seems to have lost some 'discriminative' power. Specifically, these objects are detected even when the prompt refers to something completely different.

For instance, after training on 'hole', 'crack', and 'deep crack', the model still identifies these features even when I use 'car' as a prompt.

Have you observed a similar pattern in your experiments? Are there specific training settings or hyperparameters you would recommend to mitigate this behavior?

Thank you for your reply!

**Comment by Sompote** [comment]
Thank you for your feedback. We will test and investigate this issue and tell you the results.

**Comment by Wangduo428** [comment]
> Thank you for open-sourcing this project, great work!
> 
> I have been exploring your solution to improve SAM3 on specific objects while preserving its general recognition capabilities.
> 
> However, I noticed that after fine-tuning on my dataset, the model correctly identifies the target objects compared to the original SAM, but it seems to have lost some 'discriminative' power. Specifically, these objects are detected even when the prompt refers to something completely different.
> 
> For instance, after training on 'hole', 'crack', and 'deep crack', the model still identifies these features even when I use 'car' as a prompt.
> 
> Have you observed a similar pattern in your experiments? Are there specific training settings or hyperparameters you would recommend to mitigate this behavior?
> 
> Thank you for your reply!

I have met the same problem ! Do you have any idea how to fix it? Thanks!

**Comment by Sompote** [comment]
Thank you both for the detailed report and for confirming you hit the same thing — this was a real bug, and it's now fixed on main (commit 28744e5).

What was happening

SAM3 is an open-vocabulary detector: a Hungarian matcher + presence/focal loss decides, per text prompt, which object queries fire. The training dataloaders were building positive-only queries — every prompt always had matching boxes (is_exhaustive=True, non-empty object_ids_output). The model was never shown a prompt that should return nothing, so the presence head decoupled from the text condition and learned to "always fire" on a familiar-looking image. That's exactly why prompting car still lit up the cracks/holes. Two settings made it worse: LoRA was applied to the text encoder (which distorts SAM3's text↔image alignment) and the default epoch counts overfit small datasets.

The fix (already pushed)

1. Hard-negative text queries — new num_negatives config option (default 3), in both train_sam3_lora_with_categories.py and train_sam3_lora_native.py. For each image the dataloader now adds prompts that must return zero detections, drawn from other dataset categories plus a generic out-of-domain pool (car, person, dog, …). This supplies the negative supervision that was missing and restores discrimination.
2. Text encoder frozen — apply_to_text_encoder: false across the configs, to preserve the open-vocab alignment.
3. Avoid over-training on small datasets (lower LR, fewer epochs); val loss now includes the negative prompts.

To apply

git pull
# in your config:
#   training: { num_negatives: 3 }
#   lora:     { apply_to_text_encoder: false }

Then retrain. To verify, prompt an unrelated word like "car" — it should now return nothing, while crack/hole/deep crack still segment correctly. If you have a domain where some of the default generic negatives could actually appear in your images, you can lower num_negatives or override the negative pool.

Please give it a try and let me know if the behavior improves on your datasets. Thanks again for flagging this! 🙏

**Comment by Wangduo428** [comment]
Thanks for the reply and fixing the project! I have a similar thought that negative samples may help.  I am now trying to add negative samples to my own dataset. However, the coco annotation file only indicates the objects contained in the image, which gives only positive samples. Is there any suggestion on how to do it properly? Thanks!
